### ISSUE TITLE
Windows super slow startup (55 secs) - v3.0.0rc1 always compiles opencl .cl files on app run

### ISSUE THREAD

**Issue by HochMartinez** [darktable-org/darktable#3384]
![CacheDarktable](https://user-images.githubusercontent.com/57639310/68614337-00646480-04c1-11ea-8a44-98e234bca95d.png)

**Describe the bug**
The problem I see in the debug log is that Darktable compiles the .cl files every time I open it. It can't find and load the cached compiled files because they are saved with a different name as the one Darktable expects.

This happens with all the .cl files located in: 'C:\Users\USER_NAME\AppData\Local\Microsoft\Windows\INetCache\darktable\cached_kernels_for_IntelRHDGraphics400_2019155058\'

For example:
'atrous.cl.bin' is compiled with the following name: 'atrous.cl.bin.ea42012ca03bfaa433b24a71ddf4ae34'.
The next time you launch darktable, it can't find 'atrous.cl.bin', so it compiles it again.

I deleted the cached folder, launched Darktable several times and watched what happened to the .bin files.

. The fist time you launch Darktable, it saves the .bin files with a long string appended at the end of the file, example: 'atrous.cl.bin' is saved as 'atrous.cl.bin.ea42012ca03bfaa433b24a71ddf4ae34'.

. The second time, it can't find the .bin and compiles again but this time it saves the .bin file only with the long string as name, example: 'atrous.cl.bin' is saved as 'ea42012ca03bfaa433b24a71ddf4ae34'.

. The third time it does the same as the second time, it can't find the .bin, so it compiles and again, saves the .bin files with a long number as name, example: 'atrous.cl.bin' is saved ea42012ca03bfaa433b24a71ddf4ae34'. And it replaces the previous .bin file that had this name.

**UPDATE**
..........................................
If I **rename the .bin files manually** with the correct name (without the appended string), DT doesn't find the .bin files and compiles the .cl files again.

**So .cl files are compiled, always, no matter what. And the startup time is always, 55 to 60 seconds**.
.........................................

I append a sample of the log output below.

Also, I see that even if you have opencl disabled, Darktable compiles and loads all the .cl modules, why? what for?...

Thank you and keep up the good work! :-)
...
Sample log output:

4.798949 [opencl_build_program] saving binary4.832309 [opencl_init] compiling program `atrous.cl'
4.832309 [opencl_load_program] could not load cached binary program from file `C:\Users\USER_NAME\AppData\Local\Microsoft\Windows\INetCache\darktable\cached_kernels_for_IntelRHDGraphics400_2019155058\atrous.cl.bin'! (-42)
4.838943 [opencl_load_program] could not load cached binary program, trying to compile source
4.845680 [opencl_load_program] successfully loaded program from `C:\Program Files\darktable3rc0\share\darktable\kernels\atrous.cl'
5.579130 [opencl_build_program] successfully built program
5.579130 [opencl_build_program] BUILD STATUS: 0
5.579130 BUILD LOG:
5.579130 <built-in>:1:9: warning: '__FINITE_MATH_ONLY__' macro redefined<built-in>:113:9: note: previous definition is herefcl build 1 succeeded.fcl build 2 succeeded.bcl build succeeded.
....

**To Reproduce**
Just launch Darktable.

**Expected behavior**
Darktable should load the cached .bin files.

**Platform (please complete the following information):**
. Darktable 3.0.0rc1 executed as normal user (not as administrator)
. Windows 10 updated with the latest patches
. opencl 1.2 reported in the debug log (opencl dlls in my system show 2.0.2 as version).

**Comment by junkyardsparkle** [comment]
> Also, I see that even if you have opencl disabled, Darktable compiles and loads all the .cl modules, why? what for?...

This is probably because OpenCL activation can be freely switched while darktable is running.

I don't know much about recent Windows, but InetCache does seem like a weird place to cache OpenCL kernels...

**Comment by parafin** [comment]
DLL version doesn't determine OpenCL version on their own, your hardware has to support it too. Also there's no OpenCL 2.0.2 (there's only 2 numbers in OpenCL version).

As for this bug - it's Windows-specific. darktable uses symbolic links for OpenCL cache, which isn't supported on Windows, so it fails. Someone has to re-write that part of the code to support Windows too.

**Comment by parafin** [comment]
Ah, actually there is Windows-specific code in there already. I guess it got broken at some point.

**Comment by HochMartinez** [comment]
> This is probably because OpenCL activation can be freely switched while darktable is running.

I see. Wouldn't it be more efficient to load (or unload) the modules after setting the option in the preferences...?...

> I don't know much about recent Windows, but InetCache does seem like a weird place to cache OpenCL kernels...

It is indeed...

This is the current path:
C:\Users\USER_NAME\AppData\Local\Microsoft\Windows\INetCache\darktable\cached_kernels_for_IntelRHDGraphics400_2019155063

I suppose it should be better in the 'darktable' folder in AppData. Here...
C:\Users\USER_NAME\AppData\Local\darktable\cached_kernels_for_IntelRHDGraphics400_2019155063

Or maybe in the app folder.

**Comment by HochMartinez** [comment]
> Ah, actually there is Windows-specific code in there already. I guess it got broken at some point.

That's what I thought. The log file shows windows paths.

**Comment by HochMartinez** [comment]
> DLL version doesn't determine OpenCL version on their own, your hardware has to support it too. Also there's no OpenCL 2.0.2 (there's only 2 numbers in OpenCL version).

This is the opencl.dll file info:

![openclversion](https://user-images.githubusercontent.com/57639310/68626306-678f1280-04db-11ea-870f-66067eeeac74.png)

It's not 2.0.2, my mistake. It says 2.0.2.0.

**Comment by peterbud** [comment]
I'll look into that

**Comment by HochMartinez** [comment]
> I'll look into that

Nice! Thank you, **peterbud**!

**Comment by peterbud** [comment]
So I have tried to repro this, but I could not repro on my machine with an Nvidia card. No matter how many times I have started darktable, it did not generated any additional compiled opencl binaries.

I have tried on my laptop, but unfortunately it has an Intel NEO OpenCL driver which is backlisted, so no chance to test it on that.

I'd need somebody else to help repro this, whether it is a common issue or an isolated problem.

@HochMartinez: could you pls clear the opencl cache and run the opencl-test program at least twice subsequently, and share the output? Use the following command line: `"c:\Program Files\darktable\bin\darktable-cltest.exe"`

BTW, based on the log you have share, it seems to me that darktable finds the cached file and tries to open it ([here](https://github.com/darktable-org/darktable/blob/master/src/common/opencl.c#L1661)) it but fails to read it ([here](https://github.com/darktable-org/darktable/blob/master/src/common/opencl.c#L1667)). The error message is just using the variable `binname`, which does not contain the MD5 sum, but the cache file is opened using the variable `cached` ([here](https://github.com/darktable-org/darktable/blob/master/src/common/opencl.c#L1632)), which DOES contain the MD5 sum - we might replace it to be very clear in the log if something fails.

The return value (-42) means `clCreateProgramWithBinary ` returned `CL_INVALID_BINARY`  ([here](https://github.com/KhronosGroup/OpenCL-Headers/blob/master/CL/cl.h#L211))

**Comment by HochMartinez** [comment]
Hi peterbud,

Thank you for taking your time to solve this issue.

I executed opencl-test two times. I append the output of the terminal and screenshots of the cache folder.

In my system, open-cl and Darktable take 60 seconds aprox. to compile the .cl files and load the binaries. Darktable does this every time I launch the application, which I find extremely inefficient and a test of patience.

Also as I wrote, it is not logical to compile or load all those binaries when opencl is disabled.

It would be great to compile, load or unload the modules when you select opencl in the preferences. Or maybe easier to program, force a restart of Darktable, when you check opencl, so it compiles the .cl files and loads the binaries on startup.

Thanks again and keep up the good work.

....

![Cache_SECOND_RUN](https://user-images.githubusercontent.com/57639310/68766989-38ce8480-0620-11ea-8a6f-988e6c93f050.png)

[darktable-cltest_SECOND_RUN.txt](https://github.com/darktable-org/darktable/files/3841279/darktable-cltest_SECOND_RUN.txt)

![Cache_FIRST_RUN](https://user-images.githubusercontent.com/57639310/68766994-39671b00-0620-11ea-8163-30e8c071d71c.png)

[darktable-cltest_FIRST_RUN.txt](https://github.com/darktable-org/darktable/files/3841280/darktable-cltest_FIRST_RUN.txt)

**Comment by parafin** [comment]
If you want to completely disable OpenCL in darktable you can pass --disable-opencl option to it, then it won't compile OpenCL kernels. 
Your suggestion to compile OpenCL kernels at the moment user enables OpenCL in the preferences is a bad idea - it will take the same 60 seconds then for darktable to modify a preference in GUI, which is very bad user experience. One-time long start of application is much more tolerable and expected. So please stop asking about that;)

**Comment by HochMartinez** [comment]
Parafin,

First, congratulations for your commits to the Darktable project. v3.0 is very, very impressive.
 
> If you want to completely disable OpenCL in darktable you can pass --disable-opencl option to it, then it won't compile OpenCL kernels.

Thank you parafin. I already knew that. I don't want to disable it. If you read the first message my GPU works fine with opencl. The point is that there is a bug in the windows version and it compiles the .cl files, over an over again.

> Your suggestion to compile OpenCL kernels at the moment user enables OpenCL in the >preferences is a bad idea - it will take the same 60 seconds then for darktable to modify a >preference in GUI, which is very bad user experience. One-time long start of application is much >more tolerable and expected.

I know but as you say, It is a one time compile, so maybe it is something that a user can stand.
There are two suggestions, did you read the second one?

>So please stop asking about that ;)

As you said before, I'm suggesting, not asking. Also, having contributed code for this project doesn't give you the authority to tell me what I have to do or not to do.

Please, be kind with people that try to help.

Greetings and keep up the good work! :-)

**Comment by parafin** [comment]
> > If you want to completely disable OpenCL in darktable you can pass --disable-opencl option to it, then it won't compile OpenCL kernels.
> 
> Thank you parafin. I already knew that. I don't want to disable it. If you read the first message my GPU works fine with opencl. The point is that there is a bug in the windows version and it compiles the .cl files, over an over again.
> 
> > Your suggestion to compile OpenCL kernels at the moment user enables OpenCL in the >preferences is a bad idea - it will take the same 60 seconds then for darktable to modify a >preference in GUI, which is very bad user experience. One-time long start of application is much >more tolerable and expected.
> 
> I know but as you say, It is a one time compile, so maybe it is something that a user can stand.

It is a one time compile at darktable's start too, so what's the difference?

> There are two suggestions, did you read the second one?

Requiring a restart after OpenCL is enabled can work, yes, but I don't think it will benefit anyone. The Windows compile bug should be fixed, then there is no issue, as you said, user can stand one time compile.

> 
> > So please stop asking about that ;)
> 
> As you said before, I'm suggesting, not asking. Also, having contributed code for this project doesn't give you the authority to tell me what I have to do or not to do.
> 
> Please, be kind with people that try to help.

I said please, didn't I? You're suggesting a change, I'm suggesting that it won't be implemented. None of this is binding.

**Comment by junkyardsparkle** [comment]
> It would be great to compile, load or unload the modules when you select opencl in the preferences. Or maybe easier to program, force a restart of Darktable, when you check opencl, so it compiles the .cl files and loads the binaries on startup.

I can understand why this might seem like a reasonable idea, but it turns out that for many iop rendering bugs, troubleshooting involves some toggling of that option, to narrow the source of the problem. I agree that fixing the actual problem which is reported here is much better than making OpenCL toggling more painful. :-)

**Comment by peterbud** [comment]
In order to fix the issue I should be able to reproduce it - which I cannot at the moment. 

Unless somebody is helping me identifying whether it is a widespread problem or specific to this GPU and/or driver, the only thing I can see as a help to add more diagnostic to that part of the code and see what we get there.

**Comment by HochMartinez** [comment]
Hi peterbud,

I'm a senior full stack programmer, I program in several languajes but not in c or c++.

To fix this isssue, could we add a patch to ensure the integrity of the name of the bin file?

I see two possible solutions:

1. **While compiling and saving the bin file**
After the bin compilation, Darktable could test if the file with the original name exist. If it does not exist, then try to find the file that starts with that name, example: 'atrous.cl.bin.ea42012ca03bfaa433b24a71ddf4ae34'. It could fix the name of the found file by renaming it and then load it.

2. **While loading the bin file**
If Darktable does not find the bin file, then it could try to load the bin that begins with the same name, ignoring the string appended after the .bin extension, Example: 'atrous.cl.bin.ea42012ca03bfaa433b24a71ddf4ae34', Darktable would ignore '.ea42012ca03bfaa433b24a71ddf4ae34' and rename the bin, so next time Darktable will find the bin the first time.

This one should be easy to code and test.

To test it, you could change the name of a bin file in the cache of your system, adding a dot and a string after the name, for example replace 'atrous.cl.bin' by 'atrous.cl.bin.ea42012ca03bfaa433b24a71ddf4ae34'.

What do you think? Is this possible?

Thank you!

**Comment by HochMartinez** [comment]
Hi junkyardsparkle,

> I can understand why this might seem like a reasonable idea, but it turns out that for many iop >rendering bugs, troubleshooting involves some toggling of that option, to narrow the source of the >problem. I agree that fixing the actual problem which is reported here is much better than making >OpenCL toggling more painful. :-)

I would not load modules, if opencl is unchecked in prefs. If your system doen't work with opencl, It's more time to launch and it's memory wasted. It is very noticeable in systems with low specs, like my humble atom laptop.

You could force a restart in preferences to load the bin files, **only if they are not loaded**. If they are loaded, then just used them or not, as it works now.

Just my two cents...

Keep up the good work guys, v3.0 rocks! :-)

**Comment by kit-kut** [comment]
Hi,

on my WIndows 10 Home, AMD 540X Notebook dedicated GPU and Darktable 2.6.3, I have to delete each time before starting DT C:\Users\USER_NAME\AppData\Local\Microsoft\Windows\INetCache\darktable\cached_kernels_for_* directories to enable opencl. With existing files enabling opencl fails with compilation error:

https://www.mail-archive.com/darktable-dev@lists.darktable.org/msg04909.html 

kind regards
Andreas

**Comment by chhil** [comment]
@kit-kut 
Your issue seems related to what I was seeing...

https://github.com/darktable-org/darktable/issues/3292

**Comment by junkyardsparkle** [comment]
I'm absolutely the wrong monkey to troubleshoot this (last version of Windows owned was '98), but a quick searching around gives me the impression that the INetCache directory is a "magic system folder" which hides files from non-admin users and may have special permissions settings... I could be wrong, but I have a superstitious hunch that it's really not the right place for the OpenCL cache.

**Comment by chhil** [comment]
@junkyardsparkle 
The default path is junction/symlink and can be changed by the user via the internet options on the system (warn: it will log you off  to take effect).
I don't know what win API is used by dt that writes to the inetcache.

I have started installing as admin and running dt as admin these days on my windows system.
Also, I have changed the inetcache path to a more easily accessible path on my system.

**Comment by peterbud** [comment]
> I could be wrong, but I have a superstitious hunch that it's really not the right place for the OpenCL cache.

The cache directory is not an arbitrary location, and I don't see much point arguing on this. We are using GTK and Glib, and `dt_loc_get_user_cache_dir` is initialized using Glib's `g_get_user_cache_dir `function [here](https://github.com/darktable-org/darktable/blob/master/src/common/file_location.c#L147) and that uses on Windows this directory. For more info you can see the Glib's documentation: https://developer.gnome.org/glib/stable/glib-Miscellaneous-Utility-Functions.html

So pls move forward, discussing the location is not helpful.

What I'm waiting for still is another repro of this issue with 3.0 rc0 or 3.0 rc1.

In the meantime I'll add more diagnostic to the OpenCL initialization.

**Comment by parafin** [comment]
Actually I think @junkyardsparkle might be right about location being the culprit. OS or antivirus might block access to this directory (e.g. on macOS Safari cache directory is inaccessible). The fact that GTK/glib suggests to use it like ~/.cache doesn't mean that it's actually a good idea. It's not like there's no bad design choices in those libs;)) Anyway, it should be easy to check - pass --cachedir option to darktable and point it to some normal location.

**Comment by HochMartinez** [comment]
Hi,

I changed the cache dir using _'./darktable --cachedir "C:\temp"'_ and the behaviour is the same.

. The first run, DT creates the cache directory and compiles the .cl files but **appending a unique string at the end of the name of the bin file**. See screenshot below (11:28).

. The second run, darktable compiles again the .cl files and this time, the name of the .bin file is the unique string, same as the one appended in the first run. See screenshot below (11:36).

. Third run, same as second run. It replaces the files of the second run.

If I **rename the .bin files manually** with the correct name (without the appended string), DT doen't find the .bin files and compiles the .cl files again.

**So .cl files are compiled, always, no matter what. And the startup time is always, 55 to 60 seconds**.

![temp](https://user-images.githubusercontent.com/57639310/68944043-20d83b80-07ad-11ea-9753-02a4fb102600.png)

**Comment by parafin** [comment]
OK, so loading cached OpenCL kernel fails for some reason, then it goes to this block: https://github.com/darktable-org/darktable/blob/5a6cbdfc0764cd4b2742c5a3304f5ca1be03907c/src/common/opencl.c#L1684
It should remove broken OpenCL binaries, but this part was never ported to Windows, so nothing is deleted. Then new OpenCL binaries are produced and written to files having just md5sum as name, which ordinary are renamed to the final name here: https://github.com/darktable-org/darktable/blob/5a6cbdfc0764cd4b2742c5a3304f5ca1be03907c/src/common/opencl.c#L1836
But since file with that name already exists, rename just fails (nobody bothers to check return code).
This is a bug, but it doesn't explain why loading cached OpenCL binaries fails.

**Comment by peterbud** [comment]
Thanks @parafin, that's useful, I'll fix that part.

Some other good news: I have disabled the blacklist check on my work laptop with the Intel HD Graphics card, and I have managed to repro the issue exactly as described by @HochMartinez 

On the question on why loading the OpenCL cached file fails, my current suspect is that we open the compiled file with a simple "w" [here](https://github.com/darktable-org/darktable/blob/master/src/common/opencl.c#L1818), and I think we should open it with "wb", otherwise on Windows it will do an automatic CRLF conversion with the latest glib.

Also I guess that depending on the implementation of `clCreateProgramWithBinary` with the different binary formats for the GPU drivers some of them can handle CRLF in the files  (like Nvidia), other cannot (like AMD). That might explain why some people face this issue, and others don't.

**Comment by kit-kut** [comment]
Hi,
in my case getting the cached files seems to be ok, but build fails at this point:
https://github.com/darktable-org/darktable/blob/5a6cbdfc0764cd4b2742c5a3304f5ca1be03907c/src/common/opencl.c#L1731

err is -11 after this line

**Comment by HochMartinez** [comment]
Hi guys,

@peterbud and @parafin, thank you for taking time to fix this.

>But since file with that name already exists, rename just fails (nobody bothers to check return code).
>This is a bug, **but it doesn't explain why loading cached OpenCL binaries fails**.

As I commented in a previous message, opencl is detected and **binaries are loaded**. And when opencl is enabled in the prefs, I can see gpu activity and DT feels more snappy. When it is disabled, DT definitely feels slower.

DT doesn't load the correctly named .bin files that I changed manually but It loads the ones with the checsum. I don't know if it loads the ones with the .cl name+checksum as name or the ones with the checknum as name. Probalbly the last ones it compiled.

About the path location: It is not critical but it would be nice to have one that the uninstaller is aware of. Right now the uninstaller doesn't delete the binaries cache.

Thanks and keep up the good work! :-)

**Comment by peterbud** [comment]
@HochMartinez : the correct compiled cached OpenCL binary file is which has the name `[module].cl.bin.[MD5checksum]` aka the one with the MD5 checksum in the file name at the end.

On removing the binaries cache upon uninstall: uninstaller does not do this, similar to the philosophy of not removing your darktable settings or darktable database opon uninstall. Also considering that you can  override cache and/or database directories using command line as parafin pointed out, even in case of removing the cached binaries from default location would not provide a complete solution. Now considering that the compiled binaries are taking up around ~7M space I don't think it's a major issue.

**Comment by HochMartinez** [comment]
Hi @peterbud and @parafin,

Issue is fixed.

DT windows loads the kernels ok.

And now startup is 10 secs. intead of 55 secs.

Thank you and keep up the good! :-)
